README.md
# LSM-Tree database

Handmade LSM-Tree database.

For education purposes only.

Inspired by Martin Kleppmann book:
[Designing Data-Intensive Applications: The Big Ideas Behind Reliable, Scalable, and Maintainable Systems](https://www.amazon.com/Designing-Data-Intensive-Applications-Reliable-Maintainable/dp/1449373321)


CHANGELOG.md
## 0.1.0

- Initial version, created by Ilya Lebedev <bacherko@gmail.com>


lib/merge.dart
import 'dart:convert';
import 'dart:io';
import 'dart:async';
import 'dart:isolate';
import 'dart:typed_data';
import 'dart:math';

import 'isolate_medium.dart';

const int MERGE_CHUNK_SIZE = 2048;
const int BULK_WRITE_SIZE = 8192;
const int SCHEDULER_DELAY_SEC = 10;

void runMergeScheduler(SendPort sendPort) async {
  print('Merger started');

  final medium = IsolateMedium(null, sendPort: sendPort);

  schedule(medium);
}

void schedule(IsolateMedium medium) {
  Timer(Duration(seconds: SCHEDULER_DELAY_SEC), () async {
    await run(medium);
    schedule(medium);
  });
}

class Row {
  final String key;
  final Uint8List value;

  Row(this.key, this.value);
}

class TableCursor {
  String pageName;
  RandomAccessFile file;
  Uint8List buf;
  ByteData view;
  int fileOffset = 0;
  int offset = 0;
  int chunkLength = null;

  TableCursor(this.pageName) {
    buf = Uint8List(MERGE_CHUNK_SIZE);
    view = ByteData.view(buf.buffer);
  }

  Future<void> start() async {
    file = await File('db/$pageName').open(mode: FileMode.read);
  }

  Future<bool> readNextChunk() async {
    fileOffset += offset;
    await file.setPosition(fileOffset);
    chunkLength = await file.readInto(buf);
    offset = 0;

    return chunkLength > 0;
  }

  Future<Row> readRow() async {
    if (chunkLength == 0) {
      return null;
    }

    if (chunkLength == null) {
      await readNextChunk();
    }

    if (offset + 4 > chunkLength) {
      if (!(await readNextChunk())) {
        return null;
      }
    }

    var keyLength = view.getUint16(offset);
    var valueLength = view.getUint16(offset + 2);

    if (offset + 4 + keyLength + valueLength < chunkLength) {
      await readNextChunk();
    }

    var key = String.fromCharCodes(buf, offset + 4, offset + 4 + keyLength);
    var value = buf.sublist(
        offset + 4 + keyLength, offset + 4 + keyLength + valueLength);

    offset += 4 + keyLength + valueLength;

    return Row(key, value);
  }
}

void run(IsolateMedium medium) async {
  var jsonData = await File('db/state.json').readAsString();

  var state = jsonDecode(jsonData);
  var pages = state['pages'].cast<String>();

  if (pages.length >= 2) {
    await merge(pages[0], pages[1], medium: medium);
  }
}

void merge(String pageName1, String pageName2, {IsolateMedium medium}) async {
  final List<Row> list = [];

  final t1 = TableCursor(pageName1);
  final t2 = TableCursor(pageName2);

  await Future.wait([
    t1.start(),
    t2.start(),
  ]);

  var t1value = await t1.readRow();
  var t2value = await t2.readRow();

  while (t1value != null || t2value != null) {
    if (t1value == null) {
      list.add(t2value);
      t2value = await t2.readRow();
      continue;
    }

    if (t2value == null) {
      list.add(t1value);
      t1value = await t1.readRow();
      continue;
    }

    final comparison = t1value.key.compareTo(t2value.key);

    if (comparison == 0) {
      list.add(t2value);
      t1value = await t1.readRow();
      t2value = await t2.readRow();
    } else if (comparison > 0) {
      list.add(t2value);
      t2value = await t2.readRow();
    } else {
      list.add(t1value);
      t1value = await t1.readRow();
    }
  }

  var page = Uint8List(BULK_WRITE_SIZE);
  var view = ByteData.view(page.buffer);
  var offset = 0;
  var overallOffset = 0;

  var index = Uint8List(BULK_WRITE_SIZE);
  var indexView = ByteData.view(index.buffer);
  var indexOffset = 0;

  var newPageName = 'table${Random().nextInt(4294967296)}m';

  var newPage = await File('db/$newPageName').openWrite();
  var newPageIndex = await File('db/${newPageName}_index').openWrite();

  for (var item in list) {
    var keyCodes = item.key.codeUnits;
    var keyLength = keyCodes.length;

    // Flush page data
    if (offset + 4 + keyLength + item.value.length > page.length) {
      newPage.add(page.sublist(0, offset));
      await newPage.flush();
      page = Uint8List(BULK_WRITE_SIZE);
      view = ByteData.view(page.buffer);
      offset = 0;
    }

    // Flush index data;
    if (indexOffset + 6 + keyLength > index.length) {
      newPageIndex.add(index.sublist(0, indexOffset));
      await newPageIndex.flush();
      index = Uint8List(BULK_WRITE_SIZE);
      indexView = ByteData.view(index.buffer);
      indexOffset = 0;
    }

    view.setUint16(offset, keyCodes.length);
    view.setUint16(offset + 2, item.value.length);
    page.setRange(offset + 4, offset + 4 + keyLength, keyCodes);
    page.setRange(offset + 4 + keyLength,
        offset + 4 + keyLength + item.value.length, item.value);

    indexView.setUint32(indexOffset, overallOffset);
    indexView.setUint16(indexOffset + 4, keyLength);
    index.setRange(indexOffset + 6, indexOffset + 6 + keyLength, keyCodes);

    var deltaOffset = 4 + keyLength + item.value.length;

    offset += deltaOffset;
    overallOffset += deltaOffset;
    indexOffset += 6 + keyLength;
  }

  newPage.add(page.sublist(0, offset));
  newPageIndex.add(index.sublist(0, indexOffset));

  await Future.wait([
    () async {
      await newPage.flush();
      await newPage.close();
    }(),
    () async {
      await newPageIndex.flush();
      await newPageIndex.close();
    }()
  ]);

  print('Table "${pageName1}" and "$pageName2" merged into "$newPageName"');

  await medium.call('update_state', {
    'replacePages': [pageName1, pageName2],
    'byPage': newPageName,
  });
}


lib/server.dart
import 'dart:async';
import 'dart:convert' as convert;
import 'dart:typed_data';
import 'package:shelf/shelf.dart' as shelf;
import 'package:shelf/shelf_io.dart' as io;
import 'lsm_tree.dart';

void startServer({String hostname, int port, Database db}) async {
  var handler = const shelf.Pipeline()
      // .addMiddleware(shelf.logRequests())
      .addHandler((req) => _echoRequest(req, db));

  var server = await io.serve(handler, hostname, port);
  print('Serving at http://${server.address.host}:${server.port}');
}

Future<shelf.Response> _echoRequest(shelf.Request request, Database db) async {
  var timer = Timer(Duration(seconds: 10), () {
    print('Timeout reached!');
  });

  try {
    return await __echoRequest(request, db);
  } catch (err, stack) {
    print('Request failed with error:');
    print(err);
    print(stack);
    return shelf.Response.internalServerError();
  } finally {
    timer.cancel();
  }
}

Future<shelf.Response> __echoRequest(shelf.Request request, Database db) async {
  if (request.method != 'POST') {
    return shelf.Response.forbidden('Not allowed');
  }

  if (!request.headers['content-type'].startsWith('application/json')) {
    return shelf.Response(400, body: 'Only json allowed');
  }

  var requestBody;

  try {
    requestBody = convert.jsonDecode(await request.readAsString());
  } catch (err) {
    return shelf.Response(400, body: 'Invalid body');
  }

  try {
    return await __handleApi(request.url.toString(), requestBody, db);
  } on UserParamException catch (err) {
    return shelf.Response(400, body: 'Error: ${err.cause}');
  }
}

Future<shelf.Response> __handleApi(
    String apiName, Map<String, dynamic> data, Database db) async {
  String result;

  switch (apiName) {
    case 'set':
      String key = data['key'];
      String value = data['value'];

      var bytes = Uint8List.fromList(value.codeUnits);

      await db.set(key, bytes);

      return shelf.Response.ok(convert.jsonEncode({
        'key': key,
        'value': value,
      }));
    case 'get':
      String key = data['key'];
      var value = await db.get(key);
      result = value == null ? null : String.fromCharCodes(value);

      return shelf.Response.ok(convert.jsonEncode({
        'key': key,
        'value': result,
      }));
    default:
      return shelf.Response.notFound('Method not found');
  }
}


lib/isolate_medium.dart
import 'dart:async';
import 'dart:isolate';

typedef MethodHandler = Future<dynamic> Function(
    String methodName, dynamic params);

class SendQueueItem {
  final String methodName;
  final dynamic params;
  final Completer completer = Completer();

  SendQueueItem(this.methodName, this.params);
}

class IsolateMedium {
  SendPort sendPort;
  ReceivePort receiverPort;
  final MethodHandler methodHandler;
  List<SendQueueItem> sendQueue;
  final Map<int, Completer> waits = {};
  int lastId = 0;

  IsolateMedium(this.methodHandler, {this.sendPort}) {
    receiverPort = ReceivePort();
    receiverPort.listen(listen);

    if (sendPort != null) {
      sendPort.send(receiverPort.sendPort);
    }
  }

  void listen(dynamic data) async {
    if (sendPort == null) {
      sendPort = data;

      if (sendQueue != null && sendQueue.isNotEmpty) {
        for (var item in sendQueue) {
          _send(item.methodName, item.params, item.completer);
        }
        sendQueue = null;
      }
      return;
    }

    String type = data['type'];

    switch (type) {
      case 'request':
        String methodName = data['method'];

        dynamic result;

        try {
          result = await methodHandler(methodName, data['params']);
        } catch (err, stack) {
          print('Method handle failed:');
          print(err);
          print(stack);

          sendPort.send({
            'type': 'response',
            'response_for': data['id'],
            'error': {
              'message': 'Error',
            },
          });
        }

        sendPort.send({
          'type': 'response',
          'response_for': data['id'],
          'result': result,
        });
        break;
      case 'response':
        int id = data['response_for'];

        if (data['error'] != null) {
          waits[id].complete(Future.error(data['error']));
        } else {
          waits[id].complete(data['result']);
        }
        break;
      default:
        print('Unknown event: $type');
    }
  }

  SendPort get pipe {
    return receiverPort.sendPort;
  }

  Future<dynamic> call(String methodName, dynamic params) async {
    if (sendPort == null) {
      var sendItem = SendQueueItem(methodName, params);
      sendQueue ??= [];
      sendQueue.add(sendItem);
      return sendItem.completer.future;
    }

    var completer = Completer();

    _send(methodName, params, completer);

    return completer.future;
  }

  void _send(String methodName, dynamic params, Completer completer) {
    var requestId = ++lastId;
    waits[requestId] = completer;

    sendPort.send({
      'type': 'request',
      'id': requestId,
      'method': methodName,
      'params': params,
    });
  }
}


stress-test/generate_payload.dart
import 'dart:io';
import 'dart:convert';
import 'dart:math';

void main() async {
  var writter = File('ammo.txt').openWrite(mode: FileMode.write);

  writter.write(
'''[Host: example.org]
[User-Agent: Tank]
[Content-type: application/json]
''');

  for (var i = 1; i < 100000; i++) {
    var id = Random().nextInt(20000);

    if (Random().nextInt(10) < 5) {
      var setJson = jsonEncode({
        'key': 'key$id',
        'value': 'some value of key $id by $i iteration, keks! popy',
      });

      writter.write('${setJson.length + 1} /set\n $setJson\n');
    } else {
      var getJson = jsonEncode({
        'key': 'key$id',
      });

      writter.write('${getJson.length + 1} /get\n $getJson\n');
    }

    if (i % 10000 == 0) {
      await writter.flush();
    }

  }

  await writter.flush();
  await writter.close();
}


lib/lsm_tree.dart
import 'dart:convert';
import 'dart:io';
import 'dart:async';
import 'dart:collection';
import 'dart:isolate';
import 'dart:typed_data';
import 'dart:math';
import 'package:archive/archive.dart';
import 'package:simple_bloom_filter/simple_bloom_filter.dart';
import 'isolate_medium.dart';
import 'merge.dart';

const int MEM_TABLE_SIZE_LIMIT = 5000;
const int CHUNK_SIZE = 8192;
const int MAX_KEY_LENGTH = 1000;
const int MAX_VALUE_LENGTH = 65536;

class UserParamException implements Exception {
  String cause;
  UserParamException(this.cause);
}

class Page {
  String pageName;
  bool isLoaded = false;
  Map<String, int> index = null;
  Future<void> indexLoading;
  RandomAccessFile file;
  Future currentReading = Future.value();

  Page(this.pageName);

  void close() {
    if (file != null) {
      file.close().catchError((err) {
        print('Closing failed:');
        print(err);
      });
    }
  }

  Future<void> loadIndex() async {
    if (isLoaded) {
      return;
    }

    if (indexLoading != null) {
      return indexLoading;
    }

    indexLoading = _loadIndex();
    return indexLoading;
  }

  Future<void> _loadIndex() async {
    var indexBytes = await File('db/${pageName}_index').readAsBytes();
    var view = ByteData.view(indexBytes.buffer);

    index = {};

    int offset = 0;

    while (offset < indexBytes.length) {
      var dataOffset = view.getUint32(offset);
      var keyLength = view.getUint16(offset + 4);
      var key =
          String.fromCharCodes(indexBytes, offset + 6, offset + 6 + keyLength);

      index[key] = dataOffset;

      offset += 6 + keyLength;
    }

    isLoaded = true;
    print('Index "$pageName" loaded');
  }

  Future<Uint8List> getValue(String key) async {
    var offset = index[key];

    if (offset == null) {
      return null;
    }

    var data = Uint8List(CHUNK_SIZE);

    await readFromPosition(offset, data);

    var view = ByteData.view(data.buffer);
    var keyLength = view.getUint16(0);
    var dataLength = view.getUint16(2);

    // TODO: remove
    if (keyLength > 1000) {
      print('offset: $offset');
      print('keyLength: $keyLength');
      print('dataLength: $dataLength');
    }

    var foundKey = String.fromCharCodes(data, 4, 4 + keyLength);

    if (foundKey != key) {
      throw Exception('Keys index mismatch');
    }

    if (dataLength > CHUNK_SIZE - 4 - keyLength) {
      var data = Uint8List(4 + keyLength + dataLength);

      await readFromPosition(offset, data);
    }

    return data.sublist(4 + keyLength, 4 + keyLength + dataLength);
  }

  Future<void> readFromPosition(int position, Uint8List buffer) {
    return currentReading =
        currentReading.then((_) => _readFromPosition(position, buffer));
  }

  Future<void> _readFromPosition(int position, Uint8List buffer) async {
    file ??= await File('db/$pageName').open(mode: FileMode.read);

    await file.setPosition(position);
    await file.readInto(buffer);
  }
}

class State {
  static Future<State> load() async {
    var stateFile = File('db/state.json');

    List<Page> pages = [];

    if (await stateFile.exists()) {
      var stateJson = await File('db/state.json').readAsString();
      var stateData = jsonDecode(stateJson);

      for (String pageName in stateData['pages']) {
        pages.add(Page(pageName));
      }
    } else {
      await File('db/state.json').writeAsString(jsonEncode({'pages': []}));
    }

    return State(pages);
  }

  List<Page> pages;
  Future<void> stateSaving;
  Completer stateSavingCompleter;

  State(this.pages);

  Future<void> addPage(String pageName) async {
    var page = Page(pageName);
    await page.loadIndex();
    pages.add(page);

    await _savePagesState();
  }

  Future<void> _savePagesState() async {
    Completer completer;

    if (stateSaving != null) {
      if (stateSavingCompleter == null) {
        completer = stateSavingCompleter = Completer();
        await stateSaving;
      } else {
        return await stateSavingCompleter.future;
      }
    }

    stateSaving = __savePagesState();
    await stateSaving;
    stateSaving = null;

    if (completer != null) {
      completer.complete();
      stateSavingCompleter = null;
    }
  }

  Future<void> __savePagesState() async {
    await File('db/state.json').writeAsString(jsonEncode({
      'pages': List<String>.from(pages.map((page) => page.pageName)),
    }));
  }

  Future<void> loadLastPage() async {
    if (pages.isNotEmpty) {
      await pages.last.loadIndex();
    }
  }

  Future<Uint8List> getValue(key) async {
    if (pages.isEmpty) {
      return null;
    }

    for (var page in pages.reversed) {
      if (!page.isLoaded) {
        await page.loadIndex();
      }

      var value = await page.getValue(key);

      if (value != null) {
        return value;
      }
    }

    return null;
  }

  Future<void> mergePages(List<String> replacePages, String byPage) async {
    int index;

    for (var i = 0; i < pages.length; i++) {
      var page = pages[i];

      if (page.pageName == replacePages[0]) {
        index = i;
        break;
      }
    }

    if (index == null) {
      throw Exception('Bad');
    }

    if (index == pages.length - 1) {
      throw Exception('Bad');
    }

    if (pages[index + 1].pageName != replacePages[1]) {
      throw Exception('Bad');
    }

    var page1 = pages[index];
    var page2 = pages[index + 1];

    Timer(Duration(seconds: 10), () {
      page1.close();
      page2.close();
    });

    var newPage = Page(byPage);
    pages.replaceRange(index, index + 2, [newPage]);

    await _savePagesState();
    await newPage.loadIndex();

    await Future.wait([
      File('db/${replacePages[0]}').delete(),
      File('db/${replacePages[0]}_index').delete(),
      File('db/${replacePages[1]}').delete(),
      File('db/${replacePages[1]}_index').delete(),
    ]);
  }
}

class Database {
  SplayTreeMap<String, Uint8List> memtable;
  IOSink currentLog;
  Completer pageCreatingCompleter;
  State state;
  bool _isMemTableSavingStarted = false;

  Database() {
    memtable = SplayTreeMap<String, Uint8List>();
  }

  void start() async {
    print('Starting...');

    await Directory('db').create();
    state = await State.load();

    await state.loadLastPage();

    Uint8List logData;

    try {
      logData = await File('db/log').readAsBytes();
    } catch (err) {
      // Do nothing
    }

    if (logData != null && logData.isNotEmpty) {
      applyLogData(logData);
    }

    currentLog = File('db/log').openWrite(mode: FileMode.writeOnlyAppend);

    print('Database started');

    startMergeIsolate();
  }

  void startMergeIsolate() {
    var medium = IsolateMedium((String methodName, dynamic params) async {
      switch (methodName) {
        case 'update_state':
          await state.mergePages(
            params['replacePages'].cast<String>(),
            params['byPage'],
          );
          return;
        default:
          throw Exception('Unknown method');
      }
    });

    // ignore: unawaited_futures
    Isolate.spawn(runMergeScheduler, medium.pipe).catchError((err) {
      print('Merge isolate failed:');
      print(err);
    });
  }

  void applyLogData(Uint8List logData) {
    var view = ByteData.view(logData.buffer);
    var blockOffset = 0;

    while (blockOffset < logData.length) {
      var keyLength = view.getUint16(blockOffset);
      var valueLength = view.getUint16(blockOffset + 2);

      var key = String.fromCharCodes(
          logData, blockOffset + 4, blockOffset + 4 + keyLength);
      var value = logData.sublist(blockOffset + 4 + keyLength,
          blockOffset + 4 + keyLength + valueLength);

      memtable[key] = value;

      blockOffset += 4 + keyLength + valueLength;
    }
  }

  Future<Uint8List> get(String key) async {
    return memtable[key] ?? await state.getValue(key);
  }

  void set(String key, Uint8List value) async {
    var keyCodes = key.codeUnits;

    if (keyCodes.length > MAX_KEY_LENGTH) {
      throw UserParamException('Key too big');
    }

    if (value.length > MAX_VALUE_LENGTH) {
      throw UserParamException('Value too big');
    }

    var header = Uint8List(4);
    var view = ByteData.view(header.buffer);

    view.setUint16(0, keyCodes.length);
    view.setUint16(2, value.length);

    // Wait page creation if in process now
    if (pageCreatingCompleter != null) {
      await pageCreatingCompleter.future;
    }

    currentLog.add(header);
    currentLog.add(keyCodes);
    currentLog.add(value);

    memtable[key] = value;

    checkMemTableLimit();
  }

  void checkMemTableLimit() {
    if (!_isMemTableSavingStarted && memtable.length >= MEM_TABLE_SIZE_LIMIT) {
      _isMemTableSavingStarted = true;
      Timer(Duration(milliseconds: 1), saveMemTableData);
    }
  }

  void saveMemTableData() async {
    pageCreatingCompleter = Completer();

    var pageName = 'table${Random().nextInt(4294967296)}';

    var snapshotSize = 0;
    var indexSize = 0;

    for (var key in memtable.keys) {
      var keyCodes = key.codeUnits;

      snapshotSize += 4 + keyCodes.length + memtable[key].length;
      indexSize += 6 + keyCodes.length;
    }

    var snapshot = Uint8List(snapshotSize);
    var view = ByteData.view(snapshot.buffer);
    var index = Uint8List(indexSize);
    var viewIndex = ByteData.view(index.buffer);
    var bloom = simple_bloom_filter(8192, 3);

    int offset = 0;
    int indexOffset = 0;

    for (var key in memtable.keys) {
      var value = memtable[key];
      var keyCodes = key.codeUnits;
      var keyLength = keyCodes.length;
      var valueLength = value.length;

      view.setUint16(offset, keyLength);
      view.setUint16(offset + 2, valueLength);
      snapshot.setRange(offset + 4, offset + 4 + keyLength, keyCodes);
      snapshot.setRange(
          offset + 4 + keyLength, offset + 4 + keyLength + valueLength, value);

      viewIndex.setUint32(indexOffset, offset);
      viewIndex.setUint16(indexOffset + 4, keyLength);
      index.setRange(indexOffset + 6, indexOffset + 6 + keyLength, keyCodes);

      bloom.add(key);

      offset += 4 + keyLength + valueLength;
      indexOffset += 6 + keyLength;
    }

    await File('db/$pageName').writeAsBytes(snapshot);
    await File('db/${pageName}_index').writeAsBytes(index);

//    var bytesCount = (bloom.bitArray.length / 8).ceil();
//    var bloomList = Uint8List(bytesCount);
//    bloomList.fillRange(0, bytesCount, 0);
//
//    for (var byteIndex = 0; byteIndex < bytesCount; byteIndex++) {
//      int value = 0;
//
//      for (var biteIndex = 0; biteIndex < 8; biteIndex++) {
//        if (bloom.bitArray[byteIndex * 8 + biteIndex]) {
//          value += 1 << biteIndex;
//        }
//      }
//
//      bloomList[byteIndex] = value;
//    }
//
//    await File('db/${pageName}_bloom').writeAsBytes(bloomList);

    print('Table "$pageName" created');

    await currentLog.close();
    currentLog = File('db/log').openWrite(mode: FileMode.writeOnly);
    memtable.clear();
    pageCreatingCompleter.complete();

    print('Mem table cleared');

    await state.addPage(pageName);

    _isMemTableSavingStarted = false;
  }
}

Future<int> calculate() async {
  var file = File('bin/test.txt');

  var a = await file.open(mode: FileMode.read);

  await a.setPosition(5);

  var result = await a.read(100);

  var encoded = GZipEncoder().encode(result);

  print(result.length);
  print(encoded.length);

  var decoded = GZipDecoder().decodeBytes(encoded);

  print(String.fromCharCodes(decoded));

  // print(result);

  // print(String.fromCharCodes(result));

  return 6 * 7;
}

Future<void> wait(Duration duration) {
  final completer = Completer();
  Timer(duration, completer.complete);
  return completer.future;
}
